blog/2022-03-07-exhaustive-switches.mdx

```diff
@@ -0,0 +1,183 @@
+
+Let's produce an example in literally the only context that has any meaning to me anymore: currency from A Song of Ice and Fire:
+
+
+```
+enum Coin {
+    Halfpenny,
+    Penny,
+    Halfgroat,
+    Groat,
+    Star,
+    Stag,
+    Moon,
+}
+
+function printValue(coin: Coin) {
+  switch (coin) {
+    case Coin.Halfpenny:
+      console.log("I am worth 23520 Golden Dragons")
+      break;
+    case Coin.Penny:
+        console.log("I am worth 11760 Golden Dragons")
+      break;
+    case Coin.Halfgroat:
+        console.log("I am worth 5880 Golden Dragons")
+      break;
+    case Coin.Groat:
+        console.log("I am worth 2940 Golden Dragons")
+      break;
+    case Coin.Star:
+        console.log("I am worth 1470 Golden Dragons")
+      break;
+    case Coin.Stag:
+        console.log("I am worth 210 Golden Dragons")
+      break;
+    case Coin.Moon:
+      console.log("I am worth 30 Golden Dragons")
+      break;
+  }
+}
+```
+
+I can stick any of these coins in that I like, and i'll get my value in Dragons.
+
+But wait! I've forgotten a value in my enum. I've forgotten to put in Dragons.
+If I go and do that now:
+
+```
+enum Coin {
+    Halfpenny,
+    Penny,
+    Halfgroat,
+    Groat,
+    Star,
+    Stag,
+    Moon,
+    Dragon
+}
+```
+
+and call the `printValue()` function, I don't get any print out. I've got no default.
+
+I could add a default case that handles any unaccounted-for values:
+
+```
+function printValue(coin: Coin) {
+  switch (coin) {
+    case Coin.Halfpenny:
+      console.log("I am worth 23520 Golden Dragons")
+      break;
+    case Coin.Penny:
+        console.log("I am worth 11760 Golden Dragons")
+      break;
+    case Coin.Halfgroat:
+        console.log("I am worth 5880 Golden Dragons")
+      break;
+    case Coin.Groat:
+        console.log("I am worth 2940 Golden Dragons")
+      break;
+    case Coin.Star:
+        console.log("I am worth 1470 Golden Dragons")
+      break;
+    case Coin.Stag:
+        console.log("I am worth 210 Golden Dragons")
+      break;
+    case Coin.Moon:
+      console.log("I am worth 30 Golden Dragons")
+      break;
+    default:
+      console.log("I don't recognise this currency?")
+  }
+}
+```
+
+But my default needs to be generic no matter how may new coins I add. it can't ever be a handler for custom logic new items I add to the enum.
+
+However, we can use the default case to **assign** the coin to `never`:
+
+```
+    default:
+      const exhaust: never = coin;
+      throw new Error(exhaust);
+  }
+```
+
+And now the moment we try to compile our code with an enum case unaccounted for, we get the following:
+
+
+```
+index.ts:36:13 - error TS2322: Type 'Coin' is not assignable to type 'never'.
+
+36       const exhaust: never = coin;
+~~~~~~~
+
+
+Found 1 error.
+```
+
+The typescript compiler is now actively working for us to prevent any bugs that might arise from missing branches in this switch. It is
```

Comment:
```suggestion
The Typescript compiler is now actively working for us to prevent any bugs that might arise from missing branches in this switch. It is
```